File: node4/src/runtime/agent-language.ts
/**
 * Agent output-language registry + template-based policy formatter (#134 / #135).
 *
 * Catalog source: agent-language-catalog.json (shipped copy of shared/).
 * Extending languages = edit that JSON once and sync shipped copies — not
 * per-locale inject branches or session-path edits.
 */

import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import {
  renderPromptTemplate,
  sanitizeLanguageTemplateValue,
} from "./prompt-template.js";

/** Wire codes shipped in this epic (plus auto). */
export type AgentLanguageCode = "auto" | "zh-CN" | "zh-TW" | "en" | "ja";

export type AgentLanguageRegistryEntry = {
  code: AgentLanguageCode;
  uiLabel: string;
  optionLabel: string;
  promptName?: string;
  aliases: string[];
};

type CatalogFile = {
  default: string;
  languages: Array<{
    code: string;
    ui_label: string;
    option_label: string;
    prompt_name?: string;
    aliases?: string[];
  }>;
};

function loadCatalog(): CatalogFile {
  const here = dirname(fileURLToPath(import.meta.url));
  const path = join(here, "agent-language-catalog.json");
  return JSON.parse(readFileSync(path, "utf8")) as CatalogFile;
}

const CATALOG = loadCatalog();

/**
 * Registry built from the shipped JSON catalog.
 * Adding a language later = append a JSON row + sync copies (see shared/).
 */
export const AGENT_LANGUAGE_REGISTRY: readonly AgentLanguageRegistryEntry[] =
  CATALOG.languages.map((row) => ({
    code: row.code as AgentLanguageCode,
    uiLabel: row.ui_label,
    optionLabel: row.option_label,
    promptName: row.prompt_name,
    aliases: row.aliases ?? [],
  }));

export const DEFAULT_AGENT_LANGUAGE: AgentLanguageCode =
  (CATALOG.default as AgentLanguageCode) || "auto";

export const FORCED_AGENT_LANGUAGE_CODES: readonly AgentLanguageCode[] =
  AGENT_LANGUAGE_REGISTRY.filter((e) => e.code !== "auto").map((e) => e.code);

export const AGENT_LANGUAGE_CODES: readonly AgentLanguageCode[] =
  AGENT_LANGUAGE_REGISTRY.map((e) => e.code);

const CODE_SET = new Set<string>(AGENT_LANGUAGE_CODES);

const ALIAS_TO_CODE = new Map<string, AgentLanguageCode>();
for (const entry of AGENT_LANGUAGE_REGISTRY) {
  for (const alias of entry.aliases) {
    ALIAS_TO_CODE.set(aliasKey(alias), entry.code);
  }
}

function aliasKey(raw: string): string {
  const s = raw.trim();
  if (/[\u3040-\u30ff\u3400-\u9fff]/.test(s)) return s;
  return s.toLowerCase().replace(/_/g, "-");
}

export type ResolvedAgentLanguage =
  | { mode: "auto"; code: "auto" }
  | {
      mode: "forced";
      code: Exclude<AgentLanguageCode, "auto">;
      promptName: string;
      uiLabel: string;
    };

export function getAgentLanguageEntry(
  code: string,
): AgentLanguageRegistryEntry | undefined {
  return AGENT_LANGUAGE_REGISTRY.find((e) => e.code === code);
}

/**
 * Normalize free-form operator / envelope input to a registry code.
 * Unknown values → `auto` (safe default for runtime inject).
 */
export function normalizeAgentLanguage(raw: unknown): AgentLanguageCode {
  const s = String(raw ?? "").trim();
  if (!s) return DEFAULT_AGENT_LANGUAGE;
  if (CODE_SET.has(s)) return s as AgentLanguageCode;

  const key = aliasKey(s);
  const fromAlias = ALIAS_TO_CODE.get(key);
  if (fromAlias) return fromAlias;

  for (const code of AGENT_LANGUAGE_CODES) {
    if (code.toLowerCase() === key) return code;
  }

  return DEFAULT_AGENT_LANGUAGE;
}

export function resolveAgentLanguage(raw: unknown): ResolvedAgentLanguage {
  const code = normalizeAgentLanguage(raw);
  if (code === "auto") return { mode: "auto", code: "auto" };
  const entry = getAgentLanguageEntry(code);
  if (!entry || !entry.promptName) {
    return { mode: "auto", code: "auto" };
  }
  return {
    mode: "forced",
    code: code as Exclude<AgentLanguageCode, "auto">,
    promptName: entry.promptName,
    uiLabel: entry.uiLabel,
  };
}

export const FORCED_LANGUAGE_POLICY_TEMPLATE = [
  "## Output language (node policy: {{ language_code }})",
  "Write **all agent-authored narrative** in **{{ language_prompt_name }}** when this node policy is set.",
  "Surfaces in scope: user-facing chat replies; thinking/reasoning text shown in Chat; todo/plan labels and updates you write; tool-call intent and progress narration; finding ledger fields (title, description, impact, remediation, PoC narrative); stage completion summaries and package/subagent handoff narration you write; report markdown you author.",
  "Do not default to another natural language for these surfaces — even if mission packs, tool output, or payloads are in English or another language.",
  "Keep technical tokens as-is when needed (paths, headers, CVE ids, code, shell stdout excerpts in proof).",
  "**Do not rewrite** raw tool stdout/stderr, HTTP bodies, or other protocol/machine identifiers; only your narration and booked finding/report text must follow this language.",
].join("\n");

export const AUTO_LANGUAGE_POLICY_TEMPLATE = [
  "## Output language (node policy: auto)",
  "Match the **user's language** for all agent-authored narrative (chat replies; thinking/reasoning text shown in Chat; todo/plan labels and updates you write; tool-call intent and progress narration; finding ledger fields; stage completion summaries and package/subagent handoff narration you write; report markdown you author).",
  "If the user writes Chinese, use Chinese; if English, use English; similarly follow other languages the user uses.",
  "Keep technical tokens as-is when needed (paths, headers, CVE ids, code, shell stdout excerpts in proof).",
  "**Do not rewrite** raw tool stdout/stderr, HTTP bodies, or other protocol/machine identifiers.",
].join("\n");

/**
 * Render the language policy block for a system prompt.
 * Shared by free OMP, Hard Graph stage, and subagent session builders.
 */
export function formatAgentLanguageInjection(language: unknown): string {
  const resolved = resolveAgentLanguage(language);
  if (resolved.mode === "auto") {
    return AUTO_LANGUAGE_POLICY_TEMPLATE;
  }
  return renderPromptTemplate(
    FORCED_LANGUAGE_POLICY_TEMPLATE,
    {
      language_code: resolved.code,
      language_prompt_name: resolved.promptName,
    },
    { sanitizeValue: sanitizeLanguageTemplateValue },
  );
}

export function isAgentLanguageCode(raw: unknown): raw is AgentLanguageCode {
  return typeof raw === "string" && CODE_SET.has(raw);
}

/**
 * Extract and normalize agent language from a platform task_assign or user_steer.
 * Always returns a registry wire code (unknown/missing → auto).
 */
export function extractAgentLanguageFromMessage(
  message: Record<string, unknown>,
): AgentLanguageCode {
  const limits =
    message.worker_limits &&
    typeof message.worker_limits === "object" &&
    !Array.isArray(message.worker_limits)
      ? (message.worker_limits as Record<string, unknown>)
      : message.workerLimits &&
          typeof message.workerLimits === "object" &&
          !Array.isArray(message.workerLimits)
        ? (message.workerLimits as Record<string, unknown>)
        : {};
  const raw =
    typeof message.agent_language === "string"
      ? message.agent_language
      : typeof message.agentLanguage === "string"
        ? message.agentLanguage
        : typeof limits.agent_language === "string"
          ? limits.agent_language
          : typeof limits.agentLanguage === "string"
            ? limits.agentLanguage
            : undefined;
  return normalizeAgentLanguage(raw);
}

/** Absolute path of the shipped catalog JSON (for cross-stack lock tests). */
export function agentLanguageCatalogPath(): string {
  return join(dirname(fileURLToPath(import.meta.url)), "agent-language-catalog.json");
}

// Re-export language sanitize for tests that imported from agent-language before.
export { sanitizeLanguageTemplateValue } from "./prompt-template.js";

File: platform/backend/app/services/agent_language.py
"""
Agent output-language catalog for Node config (#134 / #136).

Must stay in lockstep with:
- node4/src/runtime/agent-language.ts (AGENT_LANGUAGE_REGISTRY)
- platform/frontend/src/lib/agentLanguages.ts

Adding a language = extend SHIPPED_AGENT_LANGUAGES + aliases here,
plus Node registry and FE options. No session-path edits.
"""
from __future__ import annotations

from typing import Final

DEFAULT_AGENT_LANGUAGE: Final[str] = "auto"

# Wire codes accepted on Node config / worker_limits (including auto).
SHIPPED_AGENT_LANGUAGES: Final[tuple[str, ...]] = (
    "auto",
    "zh-CN",
    "zh-TW",
    "en",
    "ja",
)

ALLOWED_AGENT_LANGUAGES: Final[frozenset[str]] = frozenset(SHIPPED_AGENT_LANGUAGES)

# Alias (lowercased latin / as-is CJK) → canonical code.
# Exact wire codes are accepted before this map.
_ALIAS_TO_CODE: Final[dict[str, str]] = {
    # auto
    "follow": "auto",
    "match": "auto",
    "跟随用户": "auto",
    "跟随": "auto",
    # zh-CN
    "zh": "zh-CN",
    "zh-cn": "zh-CN",
    "chinese": "zh-CN",
    "simplified": "zh-CN",
    "simplified-chinese": "zh-CN",
    "simplified chinese": "zh-CN",
    "中文": "zh-CN",
    "简体": "zh-CN",
    "简体中文": "zh-CN",
    # zh-TW (never collapse into zh-CN)
    "zh-tw": "zh-TW",
    "zh-hant": "zh-TW",
    "zh-hk": "zh-TW",
    "traditional": "zh-TW",
    "traditional-chinese": "zh-TW",
    "traditional chinese": "zh-TW",
    "繁體": "zh-TW",
    "繁体": "zh-TW",
    "繁體中文": "zh-TW",
    "繁体中文": "zh-TW",
    # en
    "en-us": "en",
    "en-gb": "en",
    "english": "en",
    # ja
    "jp": "ja",
    "ja-jp": "ja",
    "japanese": "ja",
    "日本語": "ja",
}


def _alias_key(raw: str) -> str:
    s = raw.strip()
    if any("\u3040" <= ch <= "\u30ff" or "\u3400" <= ch <= "\u9fff" for ch in s):
        return s
    return s.lower().replace("_", "-")


def normalize_agent_language(value: object) -> str:
    """
    Return a shipped wire code (auto | zh-CN | zh-TW | en | ja).
    Unknown / empty → auto (safe default for worker_limits inject).
    """
    raw = str(value if value is not None else DEFAULT_AGENT_LANGUAGE).strip()
    if not raw:
        return DEFAULT_AGENT_LANGUAGE
    if raw in ALLOWED_AGENT_LANGUAGES:
        return raw
    key = _alias_key(raw)
    if key in ALLOWED_AGENT_LANGUAGES:
        return key
    # Case-insensitive match on shipped codes (e.g. JA → ja).
    for code in SHIPPED_AGENT_LANGUAGES:
        if code.lower() == key:
            return code
    mapped = _ALIAS_TO_CODE.get(key)
    if mapped:
        return mapped
    return DEFAULT_AGENT_LANGUAGE


def merge_worker_limits_into_message(
    msg: dict,
    limits: dict | None,
) -> dict:
    """
    Attach Node worker_limits (including agent_language) onto a task_assign
    or user_steer rebuild so language does not silently fall back to auto (#138).

    Pure: does not mutate the input dict.
    """
    out = dict(msg)
    if not limits:
        return out
    out["worker_limits"] = limits
    lang = limits.get("agent_language")
    if isinstance(lang, str) and lang.strip():
        # Belt-and-suspenders: top-level field for Node normalizeTask readers.
        if not out.get("agent_language") and not out.get("agentLanguage"):
            out["agent_language"] = lang.strip()
    return out


def parse_agent_language_for_update(value: object) -> str:
    """
    Normalize a PATCH body value to a shipped code.
    Raises ValueError if the input is free-form garbage (not alias/code).
    Empty string is treated as invalid (prefer explicit 400).
    """
    raw = str(value if value is not None else "").strip()
    if not raw:
        raise ValueError("agent_language must be one of: " + ", ".join(SHIPPED_AGENT_LANGUAGES))
    if raw in ALLOWED_AGENT_LANGUAGES:
        return raw
    key = _alias_key(raw)
    if key in ALLOWED_AGENT_LANGUAGES:
        # only if key is a real code form (ja, en, …) — not alias-only keys in ALLOWED
        return key
    for code in SHIPPED_AGENT_LANGUAGES:
        if code.lower() == key:
            return code
    mapped = _ALIAS_TO_CODE.get(key)
    if mapped:
        return mapped
    raise ValueError(
        "agent_language must be one of: " + ", ".join(SHIPPED_AGENT_LANGUAGES)
    )
